beluga/test/beluga/actions/test_assign.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <list>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include <beluga/actions/assign.hpp>
+
+#include <range/v3/action/drop.hpp>
+#include <range/v3/action/remove.hpp>
+#include <range/v3/algorithm/equal.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/transform.hpp>
+
+namespace {
+
+TEST(AssignAction, ViewToAction) {
+  auto input = std::vector{1, 2, 3};
+  input |= ranges::views::reverse | beluga::actions::assign;
+  ASSERT_TRUE(ranges::equal(input, std::vector{3, 2, 1}));
+}
+
+TEST(AssignAction, ViewToActionComposition) {
+  auto reverse_and_assign = ranges::views::reverse | beluga::actions::assign;
+  auto input = std::vector{1, 2, 3};
+  input |= reverse_and_assign;
+  ASSERT_TRUE(ranges::equal(input, std::vector{3, 2, 1}));
+}
+
+TEST(AssignAction, ViewToActionCall) {
+  auto input = std::vector{1, 2, 3};
+  beluga::actions::assign(input, ranges::views::reverse);
+  ASSERT_TRUE(ranges::equal(input, std::vector{3, 2, 1}));
+}
+
+TEST(AssignAction, ViewToActionOperator) {
+  auto input = std::vector{1, 2, 3};
+  input <<= ranges::views::reverse;
```

Comment:
@nahueespinosa meta: I'll use the freedom of knowing you really know your sh*t to complain a bit :smiley: 

It is not that obvious what this does on first pass. It is also somewhat limiting for this to always be an in-place operation. Can we just switch to the stream operator? Then this becomes `input << input | ranges::views::reverse;` (verbose but clearer IMO) or even `output << input | ranges::views::reverse;` if we don't want that in-place operation. I acknowledge I may be leaving the scope of range action semantics here.

Reply:
> It is not that obvious what this does on first pass.

I think this is key, I wasn't sure whether to include the operator overloading or not. Removed in https://github.com/Ekumen-OS/beluga/pull/287/commits/1c5eefd0fcfe924c61fa139bbbd0a1c4f9ce9130.

Actions are kind of always in-place operations. The way of assigning to a different range would be:

```cpp
Container output;
output = input | ranges::to<Container>;
```

Reply:
> Actions are kind of always in-place operations.

Had the impression, yeah. FWIW `ranges::to` is (intentionally?) more explicit than it needs to be.

Reply:
This paper discusses the details: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1206r1.pdf

Yeah, it seems to be intentional, adding range constructors was the original proposal for view materialization but `std::initializer_list` made it unworkable. :smile:  

```cpp
std::vector<int> foo = ....;
std::vector a{foo}; //constructs a std:vector<std::vector<int»
std::vector b(foo); //would construct a std::vector<int>
```